python/pyspark/pandas/series.py

```diff
@@ -6442,6 +6445,8 @@ def argmin(self, axis: Axis = None, skipna: bool = True) -> int:
             raise ValueError("axis can only be 0 or 'index'")
         sdf = self._internal.spark_frame.select(self.spark.column, NATURAL_ORDER_COLUMN_NAME)
         seq_col_name = verify_temp_column_name(sdf, "__distributed_sequence_column__")
+
+        cached = sdf.cache()
```

Comment:
I think we should probably add an internal util with a context manager, and fix `attach_distributed_sequence_column`'s documentation to say that we should always use them together.

The problem is that it's very easy to forget to cache/uncache when you call `attach_distributed_sequence_column`, and it's unclear where to cache/uncache especially to new developers

Reply:
hmm, the case maybe more complex:

1. in case an action is triggered and then a scalar is retuned, like `argmin`, we can explicitly persist the datasets and unpersist it after computation;
2. otherwise, (the most cases) another `DataFrame/Series` is returned, and we can not unpersist the cached datasets;

I am wondering if there is significant regression if we do not localcheckpoint or cache the internal dataset? indexing operations easily invoke `attach_distributed_sequence_column`/`attach_default_index`

1. if the dataset is small, recomputation maybe not a big deal;
2. if the dataset is large (like 1 billion rows, or/and 100+columns), it may consume too many memory to persist the internal datasets;

Reply:
Maybe we can avoid recomputation in another way:

For a big input dataframe `sdf`:

1. have a subquery that compute the size of partitions on `sdf.select(lit(0))`, in this case the original expressions in `sdf` will not be evaluated?
2. with the sizes array `Array[Long]`, generate the result dataframe just like the `zipWithIndex`

Reply:
One option is to truncate the plan by explicitly materializing the RDD. For example,

```python
DataFrame(spark._jsparkSession.internalCreateDataFrame(df._jdf.queryExecution().toRdd(), df._jdf.schema(), False), df.sparkSession)
```

or

```scala
Dataset.ofRows(spark, org.apache.spark.sql.execution.LogicalRDD(df.queryExecution.executedPlan.output, df.queryExecution.toRdd)(spark))
```

in Scala side. In this way, it can avoid triggering additional executions by `zipWithIndex`.

Reply:
cc @ueshin FYI

Reply:
>In this way, it can avoid triggering additional executions by zipWithIndex.

Reply:
`RDD.zipWithIndex` eagerly lunches a job to calculate the partition. So once the RDD is created from Spark Plan, it won't be created again. But yeah, it has the downside of truncating the plan.

Reply:
I was thinking about 3 methods:

1. [compute the partition sizes on another DataFrame](https://github.com/apache/spark/pull/38044): do not persist/checkpoint the RDD/DataFrame, create another `sdf.select(lit(true))` and compute the stats on it;
2. [implement `attach_distributed_sequence_column` with pure dataframe operations](https://github.com/apache/spark/pull/38059): do not persist/checkpoint the RDD/DataFrame, implement a new expression `WithinPartitionIncreasingID` which output 0-based index within each partition, and use a BHJ to add the start index for each partition; (it still has several test errors to investigate);
3. just add a PS option: which control the persist of `attach_distributed_sequence_column`